rfc/cart-api-v1.md

```diff
@@ -0,0 +1,179 @@
+
+#### `GET /resources/carts/{id}`
+
+Retrieve a single cart.
+
+##### Response
+
+Instead of returning `Cart` information, the response will be also returning the list of associated `CartItem`, as shown below:
+
+```javascript
+{
+    id: str, // Cart ID
+    cart_name: str,
+    default: bool,
+    items: List[CartItem] // Each CartItem object will not have `cart_id`.
+}
+```
+
+##### Default Cart
+
+The default cart is a special case and will be given back to the client ONLY IF `id` is `default` (all lowercase).
+
+* `default` is **the alias ID** of the default cart belonging to the session user.
+  * The actual ID of default cart will be returned in the response body.
+  * This ID is only used in this endpoint.
+* When the default cart is not available, the default cart will be created and return in the same request.
+
+Here is the pseudo code for the request handler.
+
+```python
+# Pseudo code: Request Handler
+def retrieve_one_cart(id: UUID):
+    # Define: bearer_token is the request bearer token
+    if id == 'default':
+        cart = cart_repository.get_default(user_id = bearer_token)
+
+        if not cart:
+            # Either this or: cart = cart_repository.create(user_id = bearer_token, cart_name = '')
+            cart = cart_repository.create(user_id = bearer_token, cart_name = '', default = True)
+    else:
+        cart = cart_repository.get(id)
+
+    return dict(id=cart.id,
+                cart_name=cart.cart_name,
+                entity_type=cart.entity_type,
+                default=cart.default,
+                items=cart_item_repository.find_many_by_cart_id(cart.id))
+```
+
+Here is the sample request.
+
+```
+GET /resources/carts/default
+Authorization: Bearer 261e1aea-82c9-4e38-a879-e2e02ef236a5
+```
+
+Here is the sample response.
+
+```javascript
+{
+    id: "6b801de6-11d4-46c0-aed1-ca2110aa7d3c"
+    cart_name: "",
+    default: true,
+    items: [...]
+}
+```
+
+#### `PUT /resources/carts/{id}`
+
+Here is the request body.
+
+```javascript
+{
+    cart_name: str // Required
+}
+```
+
+and the response is basically the JSON representation of `Cart`.
+
+#### `DELETE /resources/carts/{id}`
+
+* The endpoint will only delete the target cart if the user creates the target cart and responds with **HTTP 200**.
+  * Otherwise, the endpoint must respond with **HTTP 403**.
+* The endpoint will respond **HTTP 410** (gone) if the cart is unknown.
+
+### CRUD APIs for `CartItem`
+
+| Method | Path | Description |
+| --- | --- | --- |
+| `POST` | `/resources/carts/{cart_id}/items/` | Create a new cart item |
+| `DELETE` | `/resources/carts/{cart_id}/items/{id}` | Delete a single cart item |
+
+#### `POST /resources/carts/{cart_id}/items/`
+
+| Param Type | Param Name | Association |
+| --- | --- | --- |
+| URL Path | `cart_id` | `Cart.id` |
+
+* The request body will be similar to `CartItem` without `cart_id`.
+* The endpoint will also respond **HTTP 409** if the corresponding cart is unknown.
```

Comment:
Why do we use 409 here?  Doesn't 404 make more sense if we can't find the cart?

Reply:
Sorry... you are right on that. Please change this to 404.